cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,326 @@
+   * @throws ArtifactNotFoundException If the given artifact does not exist
+   */
+  private File fetchArtifact(ArtifactId artifactId) throws IOException, ArtifactNotFoundException {
+    Long lastModifiedTimestamp = getCurrentLastModifiedTimestamp(artifactId);
+    String namespaceId = artifactId.getNamespace();
+    ArtifactScope scope = ArtifactScope.USER;
+    // Cant use 'system' as the namespace in the request because that generates an error, the namespace doesnt matter
+    // as long as it exists. Using default because it will always be there
+    if (ArtifactScope.SYSTEM.toString().equalsIgnoreCase(namespaceId)) {
+      namespaceId = NamespaceId.DEFAULT.getEntityName();
+      scope = ArtifactScope.SYSTEM;
+    }
+
+    String url = String.format("namespaces/%s/artifacts/%s/versions/%s/download?scope=%s",
+                               namespaceId,
+                               artifactId.getArtifact(),
+                               artifactId.getVersion(),
+                               scope);
+
+    HttpURLConnection urlConn = remoteClient.openConnection(HttpMethod.GET, url);
+    try {
+      if (lastModifiedTimestamp != null) {
+        LOG.debug("Found existing local version for {} with timestamp {}", artifactId, lastModifiedTimestamp);
+
+        ZonedDateTime lastModifiedDate = ZonedDateTime
+          .ofInstant(Instant.ofEpochMilli(lastModifiedTimestamp), ZoneId.of("GMT"));
+        urlConn.setRequestProperty(HttpHeaderNames.IF_MODIFIED_SINCE, lastModifiedDate.format(
+          DateTimeFormatter.RFC_1123_DATE_TIME));
+      }
+
+      // If we get this response that means we already have the most up to date artifact
+      if (urlConn.getResponseCode() == HttpURLConnection.HTTP_NOT_MODIFIED) {
+        LOG.debug("Call to app fabric returned NOT_MODIFIED for {} with lastModifiedTimestamp of {}", artifactId,
+                  lastModifiedTimestamp);
+        File artifactJarLocation = getArtifactJarLocation(artifactId, lastModifiedTimestamp);
+        if (!artifactJarLocation.exists()) {
+          throw new RetryableException(
+            String.format("Locally cached artifact jar for %s is missing.",
+                          artifactId));
+        }
+        return artifactJarLocation;
+      }
+
+      throwIfError(urlConn, artifactId);
+
+      ZonedDateTime newModifiedDate = getLastModifiedHeader(urlConn);
+      long newTimestamp = newModifiedDate.toInstant().toEpochMilli();
+      File newLocation = getArtifactJarLocation(artifactId, newTimestamp);
+      DirUtils.mkdirs(newLocation.getParentFile());
+
+      // Download the artifact to a temporary file then atomically rename it to the final name to
+      // avoid race conditions with multiple threads.
+      File tempFile = Files.createTempFile(newLocation.getParentFile().toPath(), String.valueOf(newTimestamp), ".jar")
+        .toFile();
+      try (InputStream in = urlConn.getInputStream()) {
+        FileUtils.copyInputStreamToFile(in, tempFile);
```

Comment:
Use `Files.copy` instead and have `tempFile` of type `Path`.

Reply:
done